vyper/contracts/libs/FulfillmentApplier.vy

```diff
@@ -0,0 +1,545 @@
+    startIndex: uint256
+) -> Execution:
+    """
+    @dev Internal pure function to aggregate a group of offer items using
+         supplied directives on which component items are candidates for
+         aggregation, skipping items on orders that are not available.
+
+    @param ordersToExecute The orders to aggregate offer items from.
+    @param offerComponents An array of FulfillmentComponent structs
+                           indicating the order index and item index of each
+                           candidate offer item for aggregation.
+    @param startIndex      The initial order index to begin iteration on when
+                           searching for offer items to aggregate.
+
+    @return execution The aggregated offer items.
+    """
+    # Get the order index and item index of the offer component.
+    orderIndex: uint256 = offerComponents[startIndex].orderIndex
+    itemIndex: uint256 = offerComponents[startIndex].itemIndex
+
+    execution: Execution = Execution({
+        item: ReceivedItem({
+            itemType: 0,
+            token: ZERO_ADDRESS,
+            identifier: 0,
+            amount: 0,
+            recipient: ZERO_ADDRESS
+        }),
+        offerer: ZERO_ADDRESS,
+        conduitKey: EMPTY_BYTES32
+    })
+
+    # Declare a variable indicating whether the aggregation is invalid.
+    # Ensure that the order index is not out of range.
+    invalidFulfillment: bool = (orderIndex >= len(ordersToExecute))
+
+    assert not invalidFulfillment, "InvalidFulfillmentComponentData"
+
+    # Get the order based on offer components order index.
+    orderToExecute: OrderToExecute = ordersToExecute[orderIndex]
+    # Ensure that the item index is not out of range.
+    invalidFulfillment = invalidFulfillment or \
+        (itemIndex >= len(orderToExecute.spentItems))
+
+    assert not invalidFulfillment, "InvalidFulfillmentComponentData"
+
+    # Get the spent item based on the offer components item index.
+    offer: SpentItem = orderToExecute.spentItems[itemIndex]
+
+    # Create the Execution.
+    execution = Execution({
+        item: ReceivedItem({
+            itemType: offer.itemType,
+            token: offer.token,
+            identifier: offer.identifier,
+            amount: offer.amount,
+            recipient: msg.sender
+        }),
+        offerer: orderToExecute.offerer,
+        conduitKey: orderToExecute.conduitKey
+    })
+
+    # Zero out amount on original offerItem to indicate it is spent
+    offer.amount = 0
+
+    # Loop through the offer components, checking for validity.
+    for component in offerComponents:
+        # Get the order index and item index of the offer component.
+        orderIndex = component.orderIndex
+        itemIndex = component.itemIndex
+
+        # Ensure that the order index is not out of range.
+        invalidFulfillment = orderIndex >= len(ordersToExecute)
+        if invalidFulfillment:
+            break
```

Comment:
can just have the assertion which takes place after the loop, here, no?

Reply:
removed it entirely since it was a redundant check